core-bundle/src/String/HtmlAttributes.php

```diff
@@ -0,0 +1,248 @@
+     * tags. The output will contain a leading space i there is at least one
+     * property set: e.g. ' foo="bar" bar="42"'.
+     */
+    public function __toString(): string
+    {
+        return $this->toString();
+    }
+
+    /**
+     * Parse attributes from an attribute string like 'foo="bar" baz="42'.
+     */
+    public static function fromString(string $attributesString): self
+    {
+        $attributes = [];
+
+        $length = \strlen($attributesString);
+        $currentName = $currentValue = null;
+        $readName = $hasDelimiter = true;
+
+        for ($position = 0; $position < $length; ++$position) {
+            // Advance position until the next character is no whitespace and return that character
+            $advance = static function () use ($attributesString, $length, &$position): string {
+                $next = '';
+
+                for ($i = $position + 1; $i < $length; ++$i) {
+                    if (($next = $attributesString[$i]) !== ' ') {
+                        break;
+                    }
+                }
+
+                $position = $i - 1;
+
+                return $next;
+            };
+
+            $currentChar = $attributesString[$position];
+            $isWhitespace = ' ' === $currentChar;
+
+            // Ignore whitespaces between names and values
+            if ($isWhitespace && (($readName && null === $currentName) || (!$readName && null === $currentValue))) {
+                continue;
+            }
+
+            // Parse name
+            if ($readName) {
+                if ($isWhitespace) {
+                    if ('=' !== $advance()) {
+                        $attributes[$currentName] = null;
+                        $currentName = '';
+                    }
+                } elseif ('=' === $currentChar) {
+                    $readName = false;
+
+                    if ('"' !== $advance()) {
+                        $hasDelimiter = false;
+                    } else {
+                        ++$position;
+                    }
+                } else {
+                    $currentName .= $currentChar;
+                }
+
+                continue;
+            }
+
+            // Parse value
+            if ('"' === $currentChar || (!$hasDelimiter && $isWhitespace)) {
+                $attributes[$currentName] = $currentValue;
+
+                $currentName = $currentValue = null;
+                $readName = $hasDelimiter = true;
+            } else {
+                $currentValue .= $currentChar;
+            }
+        }
+
+        // Add remaining value
+        if (null !== $currentValue) {
+            $attributes[$currentName] = $currentValue;
+        }
+
+        return new self($attributes);
+    }
+
+    public function set(string $name, string|int|bool|\Stringable|null $value): self
+    {
+        $name = strtolower($name);
+
+        if (1 !== preg_match('/^[a-z](?:-?[a-z0-9])*$/', $name)) {
```

Comment:
```suggestion
        if (1 !== preg_match('/^[a-z](?:[_-]?[a-z0-9])*$/', $name)) {
```
It seems that unserscores are used for custom attributes too in the wild 🙈, e.g.: https://github.com/contao/contao/pull/4204

Reply:
adjusted in 4bf6a9c8750ca54ce1f3821935a9046364f57650